### ISSUE TITLE
Specify NodeSelector/Affinities for Shoot core components

### ISSUE THREAD

**Issue by jannickfahlbusch** [gardener/gardener#2462]
**How to categorize this issue?**
/area control-plane
/kind enhancement
/priority normal

**What would you like to be added**:
I'd like to be able to specify `nodeSelector`s or `affinities` for the core component that Gardener deploys into each seed so that I can control on which worker pools those components run.

**Why is this needed**:
This is needed as some workerpools might have special hardware or might have very much resources which are reserved for workload.

By letting the core components run wherever they are scheduled they might end up on hardware which is very expensive and should only be used by the intended workload.

Allowing to specify `nodeSelector`s or `affinities` would also make it possible to create a dedicated workerpool for core components.

**Comment by rfranzke** [comment]
/cc @ggaurav10 
I remember you have already helped users in the past to control to which worker pool the system components get deployed to.

**Comment by amshuman-kr** [comment]
@jannickfahlbusch Please take a look at https://github.com/gardener/kupid. It might meet your requirement. It can be deployed as a gardener extension and can inject many scheduling criteria orthogonally (without touching anything in gardener) including nodeSelector and affinities.

I am working on integrating it with gardener seeds to optional use dedicated worker pools for specific seed workloads such as etcd. But it can be used in any Kubernetes clusters including gardener shoot clusters.

**Comment by timebertt** [comment]
@jannickfahlbusch As discussed via Slack, I think the easiest and probably also cleanest solution for your usecase would be to taint the worker pools with special hardware (ref https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#example-use-cases).

Does that work for you?

**Comment by ggaurav10** [comment]
as mentioned by TIm, taints on the nodes/worker-pool and tolerations on the workload desired to run on those worker pools should do the trick in general.

**Comment by vpnachev** [comment]
The shoot owner does not control the tolerations of the system components, of course it is possible with a mutating webhook, but this is additional effort for every shoot owner.

My opinion is that every worker pool can have a boolean field `allowSystemComponents` defaulted to `true`. When it is enabled, the nodes from the worker pool can be labeled with something like `system-components.gardener.cloud/allowed=true` as there will be a validation at least one worker pool to have this enabled. 

On the other side, the system components that are not required to run on every node (coredns, vpn-shoot, calico-typha, calico-*autoscaler, ...) will have a node selector

```
system-components.gardener.cloud/allowed=true
```

**Comment by timebertt** [comment]
> The shoot owner does not control the tolerations of the system components, of course it is possible with a mutating webhook, but this is additional effort for every shoot owner.

Yes, this is correct. But to reach @jannickfahlbusch's intended goal, he doesn't need to. The solution would probably be to taint the pools with special hardware (which the system-components should not run on) and add respective tolerations to the workload that is indeed intended to run there.

**Comment by jannickfahlbusch** [comment]
Thank you for the suggestions :)

I see that I can and should do this with Taints/Tolerations on the Shoot owner side for the deployed components.

However I think this increases complexity for components that are not directly related to the workload running on the Shoot (As the Core components are deployed by Gardener). If we'd now add a taint to all our "special" worker pools we would need to add tolerations to all components that should run on them. This would mean additional effort for a lot of teams which is not directly related to their work.

As we have many different types of clusters with different types of workload adding taints and tolerations would dramatically increase the work that we'd need to do to force the core-components onto different pools.

Therefore I think having the option to specify `allowSystemComponents: false` per worker pool would be the most sustainable solution for our use-case.

**Comment by timebertt** [comment]
I see, understood. Some setting on the worker pool would indeed be helpful here 👍

**Comment by vlerenc** [comment]
@jannickfahlbusch Indeed, without adapting your workload, we don't have a solution out-of-the-box. Except for https://github.com/gardener/kupid that @amshuman-kr was proposing. We have developed it to enforce orthogonal constraints on workload without the workload knowing. We plan to use it internally to concentrate ETCDs on a dedicated ETCD-optimised worker pool and since we can't make assumptions how a seed cluster looks like, it's nothing we can bury into Gardener, but have to supply from the outside. It's a similar case here like yours.

But I also agree with @vpnachev @rfranzke @tim-ebert that a more lightweight approach would be indeed the simplest solution.

Then again, super expensive hardware sounds like super large boxes or maybe GPU boxes? Very large boxes should not suffer too much from a few MBs of system components or am I mistaken?

Finally, you have some room between our two priority classes: :-)
```
$ k get priorityclasses
NAME                      VALUE        GLOBAL-DEFAULT   AGE
system-cluster-critical   2000000000   false            25d
system-node-critical      2000001000   false            25d
```

But that again requires adapting the workload, which you don't want and if you consider it, then this was not a serious proposal and adding the tolerations would be the better approach. ;-)

**Comment by jannickfahlbusch** [comment]
I spoke with @vpnachev over Slack. I'll try to contribute the feature.

/assign

**Comment by rfranzke** [comment]
Awesome, appreciating this @jannickfahlbusch, thank you!

**Comment by jannickfahlbusch** [comment]
Here is the PR: https://github.com/gardener/gardener/pull/2480

**Comment by vpnachev** [comment]
Today I have forcefully deleted all machines of a shoot clusters, the MCM has created the respective replacement machines.
Some of the new VMs has joined very quickly the cluster, but the last few VMs needed 20+ minutes.
While I was waiting the last VMs (there was no `CSR` or `Node` resource for them) I noticed that the nodes that has already joined the cluster were not labeled with `worker.gardener.cloud/system-components: "true"` till the last machine to join the cluster. @gardener/mcm-maintainers when do the MCM apply the labels from the `machinedeployement` on the nodes?

**Comment by vlerenc** [comment]
Uh, sounds like a good catch @vpnachev.

**Comment by vpnachev** [comment]
Yeah, I think it is somehow related with another observation - I have seen MCM keeping status of machines `Pending` while the nodes were ready for 20+ minutes, it was again on a large cluster with more than 10 VMs created simultaneously. Probably the controller was busy waiting the other VMs to join the cluster.

**Comment by hardikdr** [comment]
Hm, labels are applied almost immediately the node-object joins, there is no intended wait in between. But what you said is correct, controller might be busy doing something else meanwhile.

We also observed that recently, for large clusters, mcm getting short of the worker-threads. There is a recent PR : https://github.com/gardener/machine-controller-manager/pull/491 , to increase the number of threads which should make the situation better. 

Though, the better/long-term solution soon would be to decouple the drain from the machine-reconciliation, and unblock workers to pick-up the next machine.

**Comment by prashanth26** [comment]
>  forcefully deleted all machines of a shoot clusters, the MCM has created the respective replacement machines. Some of the new VMs has joined very quickly the cluster, but the last few VMs needed 20+ minutes.

It is a nice catch. I think we have nailed down this issue here - https://github.com/gardener/machine-controller-manager/issues/490 as mentioned by hardik. We have to have made things better with increasing the worker objects allowed. However, the permanently fix this, we need to move the drain logic out of the main machine reconcilation and we shall do that once we get sometime. Hence keeping #490 still open.

> I have seen MCM keeping status of machines Pending while the nodes were ready for 20+ minutes, it was again on a large cluster with more than 10 VMs created simultaneously

This should also be related to the same issue. I have observed that draining might take too long blocking the other machine object reconciliation (joining, rejoining, labels etc.) Will fix this once OOT is done.
